crates/sp-domains/src/valued_trie_root.rs

```diff
@@ -245,11 +245,65 @@ where
     }
 }
 
+type MemoryDB<T> = memory_db::MemoryDB<
+    <T as TrieLayout>::Hash,
+    memory_db::HashKey<<T as TrieLayout>::Hash>,
+    DBValue,
+>;
+
+/// Gnenerate a proof-of-inclusion of the value at the given `input[index]`
+///
+/// Return `None` if the given `index` out of range or fail to generate the proof
+pub fn generate_proof<Layout: TrieLayout>(input: &[Vec<u8>], index: u32) -> Option<Vec<Vec<u8>>> {
+    if input.len() <= index as usize {
+        return None;
+    }
+
+    let input: Vec<_> = input
+        .iter()
+        .enumerate()
+        .map(|(i, v)| (Compact(i as u32).encode(), v))
+        .collect();
+
+    let (db, root) = {
+        let mut db = <MemoryDB<Layout>>::default();
+        let mut root = Default::default();
+        {
+            let mut trie = <TrieDBMutBuilder<Layout>>::new(&mut db, &mut root).build();
+            for (key, value) in input.iter() {
+                trie.insert(key, value).ok()?;
+            }
+        }
+        (db, root)
+    };
+
+    let key = Compact(index).encode();
+    let proof = trie_db::proof::generate_proof::<_, Layout, _, _>(&db, &root, &[key]).ok()?;
+
+    Some(proof)
+}
+
+/// Verify the proof-of-inclusion of the `data` at the `index` of the merkle root `root`
+pub fn verify_proof<Layout>(
```

Comment:
There is already a storage proof verification in the `verification` module. We should use that instead. Ideally, we should call it Storage prover that can generate proofs and verify proofs

Reply:
I have explored this direction a bit but `verify_and_get_value` doesn't work directly, while both approaches use `trie-db` under the hook, there must be some differences between generating `StorageProof` and `trie_db::proof::generate_proof` (the later is used to generating proof-of-inclusion specifically).

I will leave a TODO for now, let me if you have a better idea.

Reply:
I think it should be generated something similart to - https://github.com/subspace/subspace/blob/main/crates/pallet-domains/src/tests.rs#L829

Anyway leave a TODO and we can revisit this